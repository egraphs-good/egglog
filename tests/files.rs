use egg_smol::*;

fn test_program(program: &str, message: &str) {
    let mut egraph = EGraph::default();
    match egraph.parse_and_run_program(program) {
        Ok(msgs) => {
            for msg in msgs {
                log::info!("  {}", msg);
            }
        }
        Err(err) => panic!("{}: {err}", message),
    }
}

fn run(path: &str) {
    let _ = env_logger::builder().is_test(true).try_init();
    let program = std::fs::read_to_string(path).unwrap();
    test_program(&program, "Top level error");

    let egraph = EGraph::default();
    let program_str = egraph
        .parse_program(&program)
        .unwrap()
        .into_iter()
        .map(|x| x.to_string())
        .collect::<Vec<String>>()
        .join("\n");
    test_program(
        &program_str,
        &format!(
            "Program:\n{}\n ERROR after parse, to_string, and parse again.",
            program_str
        ),
    );
}

// include the tests generated from the build script
include!(concat!(std::env!("OUT_DIR"), "/files.rs"));
