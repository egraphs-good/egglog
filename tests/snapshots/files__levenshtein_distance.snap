---
source: tests/files.rs
expression: serialized
---
nodes:
  IntentionStr-0:
    op: IntentionStr
    children: []
    eclass: "17"
    cost: 9223372036854776000
  ExecutionStr-0:
    op: ExecutionStr
    children: []
    eclass: "22"
    cost: 9223372036854776000
  Length-12626936695586741817:
    op: Length
    children: []
    eclass: "291"
    cost: 1
  Length-0:
    op: Length
    children: []
    eclass: "172"
    cost: 1
  Length-15952540911656918845:
    op: Length
    children: []
    eclass: "126"
    cost: 1
  Length-16783941965674463102:
    op: Length
    children: []
    eclass: "126"
    cost: 1
  Length-5871781006564002453:
    op: Length
    children: []
    eclass: "126"
    cost: 1
  Length-6755155689022739364:
    op: Length
    children: []
    eclass: "236"
    cost: 1
  Length-14289738803621830331:
    op: Length
    children: []
    eclass: "236"
    cost: 1
  Length-8417957797057827878:
    op: Length
    children: []
    eclass: "212"
    cost: 1
  Length-883374682458736911:
    op: Length
    children: []
    eclass: "212"
    cost: 1
  Length-4208978898528913939:
    op: Length
    children: []
    eclass: "86"
    cost: 1
  HorseStr-0:
    op: HorseStr
    children: []
    eclass: "5"
    cost: 9223372036854776000
  i64-10080759905092916392:
    op: "8"
    children: []
    eclass: i64-10080759905092916392
    cost: 0
  Num-10080759905092916392:
    op: Num
    children:
      - i64-10080759905092916392
    eclass: "291"
    cost: 1
  i64-0:
    op: "0"
    children: []
    eclass: i64-0
    cost: 0
  Num-0:
    op: Num
    children:
      - i64-0
    eclass: "172"
    cost: 1
  i64-5871781006564002453:
    op: "1"
    children: []
    eclass: i64-5871781006564002453
    cost: 0
  Num-5871781006564002453:
    op: Num
    children:
      - i64-5871781006564002453
    eclass: "126"
    cost: 1
  i64-4208978898528913939:
    op: "7"
    children: []
    eclass: i64-4208978898528913939
    cost: 0
  Num-4208978898528913939:
    op: Num
    children:
      - i64-4208978898528913939
    eclass: "236"
    cost: 1
  i64-16783941965674463102:
    op: "6"
    children: []
    eclass: i64-16783941965674463102
    cost: 0
  Num-16783941965674463102:
    op: Num
    children:
      - i64-16783941965674463102
    eclass: "212"
    cost: 1
  i64-11743562013128004906:
    op: "2"
    children: []
    eclass: i64-11743562013128004906
    cost: 0
  Num-11743562013128004906:
    op: Num
    children:
      - i64-11743562013128004906
    eclass: "86"
    cost: 1
  i64-17615343019692007359:
    op: "3"
    children: []
    eclass: i64-17615343019692007359
    cost: 0
  Num-17615343019692007359:
    op: Num
    children:
      - i64-17615343019692007359
    eclass: "41"
    cost: 1
  i64-5040379952546458196:
    op: "4"
    children: []
    eclass: i64-5040379952546458196
    cost: 0
  Num-5040379952546458196:
    op: Num
    children:
      - i64-5040379952546458196
    eclass: "35"
    cost: 1
  i64-10912160959110460649:
    op: "5"
    children: []
    eclass: i64-10912160959110460649
    cost: 0
  Num-10912160959110460649:
    op: Num
    children:
      - i64-10912160959110460649
    eclass: "24"
    cost: 1
  Add-12513083199194630299:
    op: Add
    children:
      - Min-8555131253702428606
      - Min-5385447922955829540
    eclass: "126"
    cost: 1
  Add-11941550287383049312:
    op: Add
    children:
      - Min-7366768035319609182
      - Num-5040379952546458196
    eclass: "24"
    cost: 1
  Add-5653980253289505627:
    op: Add
    children:
      - EditDist-17918757534547077672
      - EditDist-5073083982293673030
    eclass: "291"
    cost: 1
  Add-12305188685429859683:
    op: Add
    children:
      - Length-15952540911656918845
      - Min-12575071639800866444
    eclass: "236"
    cost: 1
  Add-9343399868448031233:
    op: Add
    children:
      - Length-16783941965674463102
      - Length-5871781006564002453
    eclass: "86"
    cost: 1
  Add-14279832564112104121:
    op: Add
    children:
      - Num-5871781006564002453
      - Min-13153029961237770003
    eclass: "41"
    cost: 1
  Add-15319128131894418994:
    op: Add
    children:
      - Add-12513083199194630299
      - Num-17615343019692007359
    eclass: "35"
    cost: 1
  Add-1029389328272588663:
    op: Add
    children:
      - Min-3618698558038355718
      - Num-10912160959110460649
    eclass: "212"
    cost: 1
  Min-11700490240027734953:
    op: Min
    children:
      - EditDist-0
      - Min-8555131253702428606
      - Min-7366768035319609182
    eclass: "172"
    cost: 1
  Min-12575071639800866444:
    op: Min
    children:
      - Length-14289738803621830331
      - Length-883374682458736911
      - Length-12626936695586741817
    eclass: "212"
    cost: 1
  Min-10304266648268463668:
    op: Min
    children:
      - Length-4208978898528913939
      - Num-11743562013128004906
      - Add-9343399868448031233
    eclass: "86"
    cost: 1
  Min-5385447922955829540:
    op: Min
    children:
      - EditDist-17918757534547077672
      - Min-10304266648268463668
      - EditDist-2776679019793341463
    eclass: "172"
    cost: 1
  Min-3618698558038355718:
    op: Min
    children:
      - Length-16783941965674463102
      - Min-13153029961237770003
      - Min-12581497049426189016
    eclass: "126"
    cost: 1
  Min-8555131253702428606:
    op: Min
    children:
      - Num-5871781006564002453
      - Length-4208978898528913939
      - Add-12513083199194630299
    eclass: "126"
    cost: 1
  Min-7366768035319609182:
    op: Min
    children:
      - Min-8555131253702428606
      - Min-7366768035319609182
      - Num-11743562013128004906
    eclass: "126"
    cost: 1
  Min-17319954900913062239:
    op: Min
    children:
      - EditDist-17918757534547077672
      - EditDist-10113463934840131226
      - Add-9343399868448031233
    eclass: "172"
    cost: 1
  Min-13153029961237770003:
    op: Min
    children:
      - Min-13153029961237770003
      - Add-14279832564112104121
      - Min-12581497049426189016
    eclass: "86"
    cost: 1
  Min-12581497049426189016:
    op: Min
    children:
      - Num-11743562013128004906
      - Test1-0
      - Num-17615343019692007359
    eclass: "86"
    cost: 1
  EditDist-6695677834520422103:
    op: EditDist
    children:
      - IntentionStr-0
      - ExecutionStr-0
    eclass: "24"
    cost: 1
  EditDist-14395355118924599440:
    op: EditDist
    children: []
    eclass: "212"
    cost: 1
  EditDist-5073083982293673030:
    op: EditDist
    children: []
    eclass: "236"
    cost: 1
  EditDist-12626936695586741817:
    op: EditDist
    children: []
    eclass: "291"
    cost: 1
  EditDist-0:
    op: EditDist
    children: []
    eclass: "172"
    cost: 1
  EditDist-7889971257895353934:
    op: EditDist
    children: []
    eclass: "172"
    cost: 1
  EditDist-2776679019793341463:
    op: EditDist
    children: []
    eclass: "172"
    cost: 1
  EditDist-8523574112360596987:
    op: EditDist
    children: []
    eclass: "172"
    cost: 1
  EditDist-10113463934840131226:
    op: EditDist
    children: []
    eclass: "172"
    cost: 1
  EditDist-17918757534547077672:
    op: EditDist
    children: []
    eclass: "126"
    cost: 1
  Unwrap-13822153148692634652:
    op: Unwrap
    children:
      - EditDist-0
    eclass: i64-0
    cost: 1
  Unwrap-11587818128845965151:
    op: Unwrap
    children:
      - Add-5653980253289505627
    eclass: i64-10080759905092916392
    cost: 1
  Unwrap-1974643878682244438:
    op: Unwrap
    children:
      - Length-16783941965674463102
    eclass: i64-5871781006564002453
    cost: 1
  Unwrap-2234512020888207708:
    op: Unwrap
    children:
      - Add-12305188685429859683
    eclass: i64-4208978898528913939
    cost: 1
  Unwrap-8885720453028561764:
    op: Unwrap
    children:
      - Add-1029389328272588663
    eclass: i64-16783941965674463102
    cost: 1
  Unwrap-6911076574346317326:
    op: Unwrap
    children:
      - Add-9343399868448031233
    eclass: i64-11743562013128004906
    cost: 1
  Unwrap-935348310899929565:
    op: Unwrap
    children:
      - Add-14279832564112104121
    eclass: i64-17615343019692007359
    cost: 1
  Unwrap-2598150418935018079:
    op: Unwrap
    children:
      - Num-5040379952546458196
    eclass: i64-5040379952546458196
    cost: 1
  Unwrap-11795535641569197560:
    op: Unwrap
    children:
      - EditDist-6695677834520422103
    eclass: i64-10912160959110460649
    cost: 1
  Test1-0:
    op: Test1
    children: []
    eclass: "41"
    cost: 9223372036854776000
root_eclasses: []
class_data:
  "17":
    type: str
  "22":
    type: str
  "291":
    type: expr
  "21":
    type: str
  "172":
    type: expr
  "0":
    type: str
  "126":
    type: expr
  "9":
    type: str
  "6":
    type: str
  "1":
    type: str
  "236":
    type: expr
  "20":
    type: str
  "15":
    type: str
  "212":
    type: expr
  "14":
    type: str
  "19":
    type: str
  "86":
    type: expr
  "7":
    type: str
  "5":
    type: str
  i64-10080759905092916392:
    type: i64
  i64-0:
    type: i64
  i64-5871781006564002453:
    type: i64
  i64-4208978898528913939:
    type: i64
  i64-16783941965674463102:
    type: i64
  i64-11743562013128004906:
    type: i64
  "41":
    type: expr
  i64-17615343019692007359:
    type: i64
  "35":
    type: expr
  i64-5040379952546458196:
    type: i64
  "24":
    type: expr
  i64-10912160959110460649:
    type: i64
  "10":
    type: str
  "12":
    type: str
  "11":
    type: str
  "18":
    type: str

